feat: refactor cli interface, add 'switch' command

shrugs opened the pull request:
I think we should either

1) Use a monorepo or
2) Use a tool like [meta](https://github.com/mateodelnorte/meta) to manage all of these repos

But I only discovered meta after writing all this code, so...

tcoulter committed:
Use npm. Allow relinking command -L

tcoulter committed:
v1.3.0

shrugs committed:
feat: refactor cli interface, add 'switch' command

gnidan commented on index.js:
```diff
@@ -1,120 +1,55 @@
+  .command('link', 'Link managed packages',
+    function() {},
+    function(argv) {
+      lint({ inDir: argv.dir });
```
(typo: `link`)

gnidan commented on index.js:
```diff
@@ -1,120 +1,55 @@
+    function(argv) {
+      use({
+        inDir: argv.dir,
+        packages: argv._,
```
Trying this out, I think we need `argv._.slice(1)` here (to omit `use` from the list of pkgs)

gnidan commented on lib/buildGitUrl.js:
```diff
@@ -0,0 +1,4 @@
+
```
blank newline

gnidan commented on lib/checkout.js:
```diff
@@ -1,11 +1,53 @@
+var path = require("path");
```
Organization thought: all these git integration wrapper modules could maybe be moved to `lib/git/checkout.js` / `lib/git/urls.js` / etc.

gnidan commented on lib/constants.js:
```diff
@@ -0,0 +1,22 @@
+
```
(is it a convention to leave a blank line if the entire file is just `module.exports`?)

shrugs replied:
I do it for arbitrary reasons, but I don't think it's a standard enforced by any linter; I'll remove them ;)

gnidan commented on lib/eachPackage.js:
```diff
@@ -0,0 +1,12 @@
+  directories.forEach(function(directory) {
+    iterator(path.basename(directory), path.resolve(inDir, directory))
+  });
+}
```
How about renaming this to just `lib/packages.js` and exporting the function in an object? (I have package iteration functionality I want to add and next to this would be a nice place)

```javascript
module.exports = {
  each: function (inDir, iterator) {
     // ...
  }
};
```

gnidan commented on README.md:
````diff
@@ -11,10 +11,27 @@ $ npm install -g truffle-checkout
+
+# Build symlinks between all packages. Run this in case of repair.
+$ tc link
+```
````
Just a note - @tcoulter had originally wanted to avoid the use of subcommands, although I think they buy us a simpler implementation (as seen by your work!). Maybe we can fall-back on undefined commands, defaulting to `use`?

gnidan commented on README.md:
````diff
@@ -25,17 +42,29 @@ There are two types of selectors: recursive selectors and one-off selectors. Exa
+tc switch <org> <branch>
+```
+
+which will `git remote add`, `git fetch`, `git checkout`, `git pull`, and `git branch` for each module (if necessary/available).
````
My first thought on this is that we can avoid this command simply by using `tc use truffle-core:develop truffle-migrate:remote/feature` and just do `git remote add` and `git fetch` under the hood.

gnidan commented on index.js:
```diff
@@ -1,120 +1,55 @@
+  .command('list', 'List managed packages',
+    function() {},
+    function(argv) {
+      list({ inDir: argv.dir });
```
How's `workspace` for the name of this option?

shrugs replied:
What do you think about `workon` ala [virtualenvwrapper](http://virtualenvwrapper.readthedocs.io/en/latest/command_ref.html) ?

gnidan reviewed (changes requested):
(Requesting changes specifically for the `lint` vs `link` typo, plus the `argv._.slice(1)` issue.)

All in all, this is a fantastic effort on your part @Shrugs. I very much like the direction you've put on this project. Thank you!

Most of my comments are naming / organization suggestions, would be curious your thoughts/responses.

Over the next day or so, I'll be looking to add a `tc test` command, using your work here as a starting point, so pardon me in advance if I move some things around on you!

Also note: I've switched the base branch to `develop`. I figure we can merge into that once individual pieces are clean and finished (so @tcoulter can give the go-ahead for various features before they make it onto `master` :)

gnidan commented:
To your point about switching to `meta`: at this point, it looks like it would add extra overhead to do so. Specifically, because I'm looking to just get CI improved by the end of the week, I think that's doable with `truffle-checkout` with a bit less effort.

@tcoulter and I plan to discuss bigger plans for dev practices (including `meta`, or a monorepo), towards the end of this month.

@mateodelnorte and I were discussing this switch for Truffle briefly yesterday, so pinging you here in case you want to tell me it'd be super trivial to switch the process over to your excellent tool :)

mateodelnorte commented:
@gnidan I'd actually argue the opposite. Kind of like investing in a stock, past performance shouldn't really sway you on whatever your decision will be today. Yes, there was already work put into creating internal truffle tooling for managing git. But, you're going to have to maintain that code as well. 

The truffle project has a bit of an unusual structure. It's unusual because the truffle repo's package.json doesn't reference most of the dependencies that actually go into creating truffle. Most idiomatic node projects will reference what they use, and require them directly. And many command line based tools use things like [commander.js](https://github.com/tj/commander.js/) to do that. But, even if you want to keep the current structure build system you all are using, you can still make things a bit more idiomatic using `meta` or https://lernajs.io/ - which plenty of popular libraries use. 

So, what are you gonna gain by switching to one of those two? Well, if all of your dependencies get pulled in by a single `npm i`, then CI building and testing gets ridiculously simple. You can have a Travis setup up and running in 5 mins. In addition, the more idiomatic your project is - the more it looks like other multi-project, open source libraries - the more tooling you're going to find magically available to you, like [semantic release](https://www.npmjs.com/package/semantic-release). 

The fork over at https://github.com/mateodelnorte/truffle should already allow you to do your build use cases. I think the only thing that would change would be your development workflow. Long term, a development workflow that looks more like other projects' is going to save you in time maintaining your own tools and allow you to build more of the stuff that matters for truffle. 

Side note, but I also I don't know if you get a benefit from building and obfuscating your code. This is all server side, public, open source code. The more easily people can read it, understand it, and modify it - the more people you'll end up getting to help you for free.

Anywhoo, that's my 2 cents! 😄

mateodelnorte commented:
By the way, regardless of the decision, happy to help out with any conversation or even lending a hand in the future!

gnidan committed:
Rename inDir to workspace

gnidan committed:
Move eachPackage to packages.forEachIn

gnidan committed:
Add utility functions for packages/dependencies

- `packages.infoFor` to retrieve a package.json
- `packages.dependenciesOf` and `packages.dependingOn` to list
  an individual package's dependers/dependees
- `packages.traverse` to recursively identify package
  dependers/dependees

gnidan committed:
Add test command

gnidan committed:
Fix incorrect variable names

gnidan committed:
Add --strict-children option

gnidan commented:
@Shrugs FYI I made a bunch of changes on branch `tc-test` ([compare](https://github.com/trufflesuite/truffle-checkout/compare/165068b...tc-test))

Shrugs commented:
@gnidan nice, I'll include those commits in this PR; I'm working on these changes now so I'll have a PR out in a little bit.

shrugs committed:
feat: PR fixes

gnidan commented:
Ah cool. Feel free to include or exclude the `tc test`-specific commits from that branch.

I am vaguely spending some time this Saturday trying to get truffle-contract working with `tc test` for CI (https://github.com/trufflesuite/truffle-contract/pull/48). Should be around to pull in your changes as you make them, or provide thoughts.

gnidan commented:
Also, @mateodelnorte and I had a brief conversation yesterday about how `meta` can help out the Truffle dev process. My current plan is to get CI working for test, not publishing (as hackily as necessary) this weekend, then revisit usage of `meta` towards the end of the month, after a features sprint.

shrugs committed:
fix: missed some rebase conflicts -_-

Shrugs commented:
After working on this for a while, I'd support using a tool that already has most/all of the features truffle needs, and I suspect lerna and meta are well suited to this. I agree with @mateodelnorte about idomatic-ness; there're a few different places where this project could benefit from adopting some standards.

I haven't extensively tested the code in this branch after merging those commits. I was running into issues with the symlinks, so I switched it to use `npm link`. There's still an issue with running `tc init` and then `tc truffle:develop`; the links make `npm install` fail for some reason.

I tried `tc test truffle-contract` (nice stuff!) but my tests are failing cause some modules aren't installed in that repo, but it should have run `npm install` beforehand...

I don't have time right now to debug this at the moment (employment lol), but I'll be back maybe tomorrow or the next day to wrap it up if it still needs doing.

gnidan commented:
> I was running into issues with the symlinks, ... 

In setting up a docker image for testing, I noticed weird `UNMET DEPENDENCY` errors in `npm list` using npm v5, along with general weirdness (maybe unrelated), so I switched back to the default npm that comes with 6.9.1, which seemed to help with weirdness. Not sure what that's about.

> so I switched it to use `npm link`.

There's a pretty strong reason why we're not using `npm link` to begin with, which is that it uses a global registry (`npm link <pkg>` uses a pre-established link for that package; you have to first do `npm link` in the <pkg> directory). This is kind of annoying with multiple `tc` workspaces, although maybe not a huge deal.

No worries, thanks for all your help!

Shrugs commented:
@gnidan you're totally right about the `npm link` global registry; in that case, using symlinks seems reasonable... how does lerna handle this part of the process?

mateodelnorte commented:
Pretty sure Lerna does symlinks. Meta just wraps npm and yarn link
commands, but I'd be happy to add an ability to use symlinks only.

On Aug 5, 2017 6:47 PM, "Matt Condon" <notifications@github.com> wrote:

> @gnidan <https://github.com/gnidan> you're totally right about the npm
> link global registry; in that case, using symlinks seems reasonable...
> how does lerna handle this part of the process?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/trufflesuite/truffle-checkout/pull/5#issuecomment-320474643>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAh3X_e7yBVHS-qQ6c0V6kzCmAFQxsfjks5sVPEMgaJpZM4OsNa->
> .
>

gnidan commented:
**Update on this:** I've gone ahead and begun to migrate processes to use `meta` for managing Truffle development.

I've created a [meta-pkgs](https://github.com/trufflesuite/meta-pkgs) plugin that replaces `tc`'s behavior.

It's a bit rough around the edges so far, but I did address the use case in this PR to be able to check out forks.

To checkout everything and get everything symlinked:

    npm install -g meta
    git clone git@github.com:trufflesuite/truffle.git
    cd truffle
    npm install
    meta git update
    meta npm install
    meta npm symlink

To checkout with `tc` syntax:

    meta pkgs checkout truffle:develop truffle-artifactor@feature

(`:` still means recursively checkout, `@` is repo only)

If you want to use a fork, first ensure the remote exists for each necessary repo:

    cd dependencies/truffle-core
    git remote add Shrugs git@github.com:Shrugs/truffle-core.git
    git fetch Shrugs
    cd ../..
    meta pkgs checkout truffle:develop truffle-core:Shrugs/develop

This is all kind of verbose right now, but hopefully a step in the right direction as far as maintainability goes.

cc @Shrugs @mateodelnorte

gnidan commented:
(Closing this PR)
